### PR TITLE
Add Travis-CI configuration

### PR DESCRIPTION
**Pull request by davidebeatrici**
Linux:
- Qt 4 on Ubuntu 14.04 (64 bit)
- Qt 5 on Ubuntu 14.04 (64 bit)

Windows:
- Qt 5 on Ubuntu 14.04 (64 bit), MinGW i686 (MXE)
- Qt 5 on Ubuntu 14.04 (64 bit), MinGW x86_64 (MXE)

### CODE REVIEW

**Comment by Kissaki** [comment]
Ubuntu 14.04 64 bit fails because it is Qt4, not 5, and QStandardPaths was introduced in 5.

We should be able to use 16.04? But is the PAHelper supposed to be Qt5 only?

https://wiki.qt.io/Transition_from_Qt_4.x_to_Qt5#QDesktopServices::storageLocation_deprecated

**Comment by davidebeatrici** [comment]
Travis-CI doesn't provide builders with Ubuntu 16.04 yet.

MumblePAHelper is not supposed to be Qt 5 only. We could use QDesktopServices instead of QStandardPaths for Qt 4.

File: .travis.yml [hunk 1]
```diff
@@ -0,0 +1,15 @@
+sudo: required # Required for apt-get build-dep
+dist: trusty
+language: cpp
+compiler: gcc
+
+env:
+  - MUMBLE_QT=qt5 MUMBLE_HOST=x86_64-linux-gnu
```
Comment (Kissaki) [inline .travis.yml]: Why does this say qt5? The dist is trusty, which is on Qt4?

**Comment by Kissaki** [comment]
~~The code issues could have been handled in a separate PR. Then the changes would be specific to travis, or what is necessary additionally for the builders.~~ I guess they actually are specific to compilers - necessary includes and init syntax.

File: Plugins.cpp [hunk 2]
```diff
@@ -31,6 +31,9 @@
 
 #include "Plugins.h"
 
+#include <QDebug>
+#include <QStandardPaths>
```
Comment (Kissaki) [inline Plugins.cpp]: Qt5 file

File: .travis.yml (hunk 1 above)
Comment (davidebeatrici) [reply .travis.yml]: Also Qt 5 is available, the problem is that it is a version prior to 5.4, which added _QStandardPaths::AppDataLocation_.

Comment (Kissaki) [reply .travis.yml]: Nvm, I see the before install script uses these envs to install stuff.

~~Why does it not work then?~~

**Comment by Kissaki** [comment]
We should add a Qt4 configuration then.

**Comment by Kissaki** [comment]
[build failed](https://travis-ci.org/mumble-voip/mumble-pahelper/jobs/233638276)
```
$ ./travis-ci/before_install.bash
[…]
$ ./travis-ci/script.bash
+'[' qt5 == qt5 ']'
+'[' x86_64-linux-gnu == x86_64-linux-gnu ']'
+git clone https://github.com/mumble-voip/mumble.git ../mumble
[…]
+qmake -Wall 'QMAKE_CXXFLAGS += -Werror'
WARNING: Operator=(QMAKE_LIBDIR_X11) clears variables previously set: /usr/share/qt4/mkspecs/default/qmake.conf:23
WARNING: Operator=(QMAKE_LIBDIR_OPENGL) clears variables previously set: /usr/share/qt4/mkspecs/default/qmake.conf:24
WARNING: Operator=(QT) clears variables previously set: /usr/share/qt4/mkspecs/features/qt.prf:54
+make -j2
/usr/lib/x86_64-linux-gnu/qt4/bin/uic mumblepahelper.ui -o release/.ui/ui_mumblepahelper.h
g++ -c -m64 -pipe -Werror -O2 -Wall -W -D_REENTRANT -DQT_NO_DEBUG -DQT_GUI_LIB -DQT_CORE_LIB -DQT_SHARED -I/usr/share/qt4/mkspecs/linux-g++-64 -I. -I/usr/include/qt4/QtCore -I/usr/include/qt4/QtGui -I/usr/include/qt4 -Irelease/.moc -Irelease/.ui -o release/.obj/Plugins.o Plugins.cpp
/usr/lib/x86_64-linux-gnu/qt4/bin/moc -DQT_NO_DEBUG -DQT_GUI_LIB -DQT_CORE_LIB -DQT_SHARED -I/usr/share/qt4/mkspecs/linux-g++-64 -I. -I/usr/include/qt4/QtCore -I/usr/include/qt4/QtGui -I/usr/include/qt4 -Irelease/.moc -Irelease/.ui Plugins.h -o release/.moc/moc_Plugins.cpp
g++ -c -m64 -pipe -Werror -O2 -Wall -W -D_REENTRANT -DQT_NO_DEBUG -DQT_GUI_LIB -DQT_CORE_LIB -DQT_SHARED -I/usr/share/qt4/mkspecs/linux-g++-64 -I. -I/usr/include/qt4/QtCore -I/usr/include/qt4/QtGui -I/usr/include/qt4 -Irelease/.moc -Irelease/.ui -o release/.obj/main.o main.cpp
Plugins.cpp:41:26: fatal error: QStandardPaths: No such file or directory
 #include <QStandardPaths>
                          ^
```

QStandardPaths should be part of Qt5 Core, which is installed by before_install.bash (libqt5core5a).

On my local Windows env, the file `QStandardPaths` includes `qstandardpaths.h`. Core is included in QT in the project file. So everything seems fine for Qt5? Why does it not find the file then?

http://doc.qt.io/qt-5/qstandardpaths.html

**Comment by davidebeatrici** [comment]
It's using Qt 4 instead of Qt 5, but it should include `QDesktopServices` and not `QStandardPaths`.

**Comment by Kissaki** [comment]
What makes you think it uses Qt4?
The travis script defines `MUMBLE_QT=qt5`, and `before_install.bash` installs Qt5 packages!?

**Comment by Kissaki** [comment]
Is Qt4 installed by default? Or how would it use Qt4 if we only install Qt5?

**Comment by davidebeatrici** [comment]
Look at the binaries path prefix: `/usr/lib/x86_64-linux-gnu/qt4/`

Yes, I think that it's installed by default.

**Comment by davidebeatrici** [comment]
For some reason the Qt version check wasn't working with Qt 4, but with Qt 5 it is.
As you can see it's including `QDesktopServices` instead of `QStandardPaths`, as Ubuntu 14.04 has Qt 5.2.1.

**Comment by Kissaki** [comment]
This PR adds an inclusion of <QDebug> in Plugins.cpp, yet the build now fails with
```
Plugins.cpp:148:48: error: ‘qInfo’ was not declared in this scope
    qInfo("Plugin: %s", qPrintable(pi->filename));
                                                ^
```

**Comment by davidebeatrici** [comment]
qInfo() was introduced in Qt 5.5.

**Comment by Kissaki** [comment]
http://doc.qt.io/qt-5/qtglobal.html#qInfo

Indeed

**Comment by Kissaki** [comment]
Plugins.cpp:91:62: error: ‘DataLocation’ is not a member of ‘QDesktopServices’

Yeah, we need to add a case for Qt >= 5.0 and < 5.4… At this point, worth creating a class/function for it that hides the preprocessor 3-case implementation logic.

**Review by mkrautz — COMMENTED** [review]
This seems fine, but why are 

> Plugins.cpp: use QSettings instead of QStandardPaths

and

> main.cpp: don't set application's name to Mumble

in this PR?

**Comment by davidebeatrici** [comment]
Ubuntu 14.04 has Qt 5.2.1, which doesn't have `QStandardPaths:AppDataLocation`.
We have to use QSettings instead.

Setting the application name was required for QStandardPaths.

**Comment by mkrautz** [comment]
Please add those important passages to the respective commit messages, then.

**Comment by davidebeatrici** [comment]
Sorry, I didn't explain in https://github.com/mumble-voip/mumble-pahelper/commit/6fd23330ad73073ac2ccb498183bd3f014ffcecc's description.

https://github.com/mumble-voip/mumble-pahelper/commit/4de277411632a03b92a8ac248f5311ce026a8595 already has the explanation in its description.

**Comment by mkrautz** [comment]
Only if you already know what's going on.

**Comment by davidebeatrici** [comment]
Yes, I meant that I explained the reason of https://github.com/mumble-voip/mumble-pahelper/commit/4de277411632a03b92a8ac248f5311ce026a8595 and not of https://github.com/mumble-voip/mumble-pahelper/commit/6fd23330ad73073ac2ccb498183bd3f014ffcecc's.

It wasn't clear in my message, sorry.

**Comment by davidebeatrici** [comment]
I also added a description to the other commits.

**Comment by davidebeatrici** [comment]
I just realized that on Linux the path retrieved from QSettings is not ".local/share/Mumble", but ".config/Mumble".

We still have to use QStandardPaths.

**Comment by davidebeatrici** [comment]
Using QStandardPaths::HomeLocation (Qt 5) or QDesktopServices::HomeLocation (Qt 4) and then manually appending the rest of the path seems to be the best solution we can use.

Regarding the Travis CI script: should we use our MXE mirror?

**Comment by mkrautz** [comment]
> Regarding the Travis CI script: should we use our MXE mirror?

Yes.

**Comment by davidebeatrici** [comment]
Done.

**Comment by hacst** [comment]
Any specific reason why we try to be backwards compatible with Qt 4 and C++98 with pa-helper? Qt 4 and anything older than Ubuntu 16.04 seems like a waste of time to me for a tool used solely for development.

**Comment by davidebeatrici** [comment]
Well... Mumble still supports Qt 4 and C++98, it would be weird to see that MumblePAHelper doesn't.

Also, it takes only a few lines of code to make the code compatible with them.

File: Plugins.cpp [hunk 3]
```diff
@@ -52,6 +58,10 @@ PluginInfo::PluginInfo() {
 }
 
 Plugins::Plugins(QObject *p) : QObject(p) {
+	// Enable all directories by default
+	bUseCurrentDirPlugins = true;
+	bUseSystemPlugins = true;
+	bUseUserPlugins = true;
```
Comment (Kissaki) [inline Plugins.cpp]: I would prefer member variables to be initialized with a [member initialization list](http://en.cppreference.com/w/cpp/language/initializer_list) - which is the equivalent to what we used before.

Comment (davidebeatrici) [reply Plugins.cpp]: Done.

**Review by Kissaki — APPROVED** [review]

**Pull request merged**